shimmy/meltingpot_compatibility.py

```diff
@@ -0,0 +1,193 @@
+"""Wrapper to convert a meltingpot substrate into a pettingzoo compatible environment
+
+Taken from
+https://github.com/deepmind/meltingpot/blob/main/examples/pettingzoo/utils.py
+and modified to modern pettingzoo API
+"""
+from __future__ import annotations
+
+import functools
+from typing import Tuple, Dict, Optional
+
+import gymnasium
+import numpy as np
+from gymnasium.utils import EzPickle
+import matplotlib.pyplot as plt
+from ml_collections import config_dict
+from pettingzoo.utils.env import ParallelEnv, AgentID
+from pettingzoo.utils.env import ObsDict, ActionDict
+
+import shimmy.utils.meltingpot as utils
+from meltingpot.python import substrate
+
+
+class MeltingPotCompatibilityV0(ParallelEnv, EzPickle):
+    """This compatibility wrapper converts a meltingpot substrate into a pettingzoo environment.
+
+    Melting Pot is a research tool developed to facilitate work on multi-agent artificial intelligence.
+    It assesses generalization to novel social situations involving both familiar and unfamiliar individuals,
+    and has been designed to test a broad range of social interactions such as: cooperation, competition,
+    deception, reciprocation, trust, stubbornness and so on.
+    Melting Pot offers researchers a set of over 50 multi-agent reinforcement learning substrates (multi-agent games)
+    on which to train agents, and over 256 unique test scenarios on which to evaluate these trained agents.
+    """
+
+    metadata = {'render_modes': ['human', 'rgb_array']}
+
+    PLAYER_STR_FORMAT = 'player_{index}'
+    MAX_CYCLES = 1000
+
+    def __init__(
+            self,
+            substrate_name: str,
+            render_mode: str | None,
+            max_cycles: int = MAX_CYCLES
+    ):
+        """Wrapper that converts a openspiel environment into a pettingzoo environment.
+
+        Args:
+            substrate_name (str): name of meltingpot substrate to load
+            render_mode (Optional[str]): render_mode
+            max_cycles (Optional[int]): maximum number of cycles (steps) before termination
+        """
+        # Load substrate from pickle
+        self.substrate_name = substrate_name
+        self.render_mode = render_mode
+        self.player_roles = substrate.get_config(self.substrate_name).default_player_roles
+        self.max_cycles = max_cycles
+        self.env_config = {"substrate": self.substrate_name, "roles": self.player_roles}
+        EzPickle.__init__(self, self.render_mode, self.env_config, self.max_cycles)
+
+        # Build substrate
+        self.env_config = config_dict.ConfigDict(self.env_config)
+        self._env = substrate.build(self.env_config['substrate'], roles=self.env_config['roles'])
+
+        self.state_space = utils.spec_to_space(
+            self._env.observation_spec()[0]['WORLD.RGB'])  # type: ignore
+
+        # Set agents
+        self._num_players = len(self._env.observation_spec())
```

Comment:
This was code from their previous pettingzoo code and the number of players stayed consistent with the observation and such, but I’ll check again to see if it might be an issue.

Reply:
What it an issue ?